rust/aptos-indexer-transaction-stream/src/utils/counters.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use prometheus::{
+    register_gauge_vec, register_int_counter_vec, register_int_gauge_vec, GaugeVec, IntCounterVec,
+    IntGaugeVec,
+};
+
+pub enum ProcessorStep {
+    ReceivedTxnsFromGrpc,
+    // Received transactions from GRPC. Sending transactions to channel.
+    ProcessedBatch,
+    // Processor finished processing one batch of transaction
+    ProcessedMultipleBatches, // Processor finished processing multiple batches of transactions
+}
+
+impl ProcessorStep {
+    pub fn get_step(&self) -> &'static str {
+        match self {
+            ProcessorStep::ReceivedTxnsFromGrpc => "1",
+            ProcessorStep::ProcessedBatch => "2",
+            ProcessorStep::ProcessedMultipleBatches => "3",
+        }
+    }
+
+    pub fn get_label(&self) -> &'static str {
+        match self {
+            ProcessorStep::ReceivedTxnsFromGrpc => {
+                "[Parser] Received transactions from GRPC. Sending transactions to channel."
+            },
+            ProcessorStep::ProcessedBatch => {
+                "[Parser] Processor finished processing one batch of transaction"
+            },
+            ProcessorStep::ProcessedMultipleBatches => {
+                "[Parser] Processor finished processing multiple batches of transactions"
+            },
+        }
+    }
+}
+
+/// These metrics are temporary (suffixed with _temp to avoid conflict with metrics in processor crate)
```

Comment:
Does it matter if they clash, don't they still mean the same thing?

Reply:
It will try to register the metric with the same name and will error. I can't import the metrics from processor crate because of cyclic dep.

Reply:
I suggest renaming them and changing them- ideally each crate is responsible for emitting its own metrics; if anything, we can allow for a "metrics_prefix" and then allow dynamically creating them (instead of a single global lazy static) per instance of the streamer. 

We can then create nice and specific metrics- maybe an "items received", and then a "filtered" either book or separate series, etc

Ie there should not be any overlap between crates and the SDK because the SDK will theoretically never have the same context on what's happening inside a given step as the step does (especially for complicated things like this)

Reply:
Ie we should move away from thinking "this is the processor GRPC client" and think of it as "this is the way anyone in any application, even python, should use our GRPC APIs, and will put it on crates.io and it'll go viral and the world will love us"- I find that generalized approach helps create nice, hermetic, batteries-included crates, vs crates that rely on specific behavior from other crates

Reply:
Also duplicate metrics- ie the GRPC emitting channel size and also processor with the instrumented channel- it's fine :-)

Reply:
Got it, didn't realize you'd be using this crate alongside the old code and the registration itself would clash.

These clients support setting a prefix. If we don't do that today, I agree we should. Example from API Gateway: https://github.com/aptos-labs/api-gateway/blob/fa335677d1c80de9dc103396d563d4e47006a661/src/metrics/labels.rs#L37.